database/query_builder.go

```diff
@@ -0,0 +1,260 @@
+package database
+
+import (
+	"fmt"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// QueryBuilder is an addon for the DB type that takes care of all the database statement building shenanigans.
+// Note: This type is designed primarily for one-off use (monouso) and subsequent disposal and should only be
+// used to generate a single database query type.
+type QueryBuilder struct {
+	subject         any
+	columns         []string
+	excludedColumns []string
+
+	// Indicates whether the generated columns should be sorted in ascending order before generating the
+	// actual statements. This is intended for unit tests only and shouldn't be necessary for production code.
+	sort bool
+}
+
+// NewQB returns a fully initialized *QueryBuilder instance for the given subject/struct.
+func NewQB(subject any) *QueryBuilder {
+	return &QueryBuilder{subject: subject}
+}
```

Comment:
I wouldn't abbreviate it here. `NewQueryBuilder` still has a reasonable length for a function name.

Also, doesn't pretty much every use of it require access to `*DB` in the end? So another option could be `func (db *DB) QueryBuilder(subject any) *QueryBuilder`.

Reply:
> I wouldn't abbreviate it here. `NewQueryBuilder` still has a reasonable length for a function name.

I would second this.

> Also, doesn't pretty much every use of it require access to `*DB` in the end? So another option could be `func (db *DB) QueryBuilder(subject any) *QueryBuilder`.

Even as this makes sense for most practical use cases, I would prefer to separate a query builder (which only builds queries, by its name) from a database connection. This also eases testing and potential mocking.

Reply:
> Even as this makes sense for most practical use cases, I would prefer to separate a query builder (which only builds queries, by its name) from a database connection. This also eases testing and potential mocking. Or, in other words, keep things as they currently are.

The struct type `QueryBuilder` remains, so that can still be used in tests directly. It could also have a member of a type like `interface { DriverName() string }` (that should be everything it needs) and the method on `*DB` would just be a convenience function for normal use. Tests could then initialize it with a dummy type that implements the same interface.